### PR TITLE
Calico

### PR DESCRIPTION
**Pull request by tomdee**
Things to do before merge
- [ ] Pin versions
  - [x] k8s policy agent docker image - do a release first.
  - [x] calico-node docker image - is v0.18.0 good enough?
  - [x] calico cni plugin - need a release with logging and hostname fixes in.
  - [ ] in built CNI plugins - need a release for host-local and flannel plugins (https://github.com/appc/cni/issues/85) . 
    - [x] Note - Need an k8s fix to tell host-local not to barf on unknown options. OR just strip them in the calico plugin...
- [x] rebase and retest with k8s 1.2
  - [ ] AWS
  - [ ] Vagrant (single too?)
  - [ ] Ensure conformance tests pass on this branch
- [ ] Documentation
  - [ ] Add policy to manual setup guide - https://github.com/coreos/coreos-kubernetes/blob/master/Documentation/getting-started.md
  - [ ] Update networking section in https://github.com/coreos/coreos-kubernetes/blob/master/Documentation/kubernetes-on-baremetal.md

- [x] Load kernel modules differently - https://coreos.com/os/docs/latest/other-settings.html And update docs.

### CODE REVIEW

File: Documentation/deploy-workers.md [hunk 1]
```diff
@@ -182,6 +199,37 @@ contexts:
...
+ +[Service]
+ +# TODO - CALICO_NETWORKING will probably change name and needs to be configurable
+ +# TODO - FIX THE RELEASE This release might support the ETCD_ENDPOINTS variable or K8s policy
+ +# TODO - Lets run under rkt?
+ +# TODO - Does HOSTNAME need to be set? What value to use?
+ +# TODO - Does /var/run/calico need to be pre-created (yes - probably under rkt)
+ +ExecStartPre=-/usr/bin/docker kill calico-node
+ +ExecStartPre=-/usr/bin/docker rm calico-node
+ +ExecStartPre=/usr/bin/docker pull calico/node:latest
+ +ExecStart=/usr/bin/docker run --net=host --privileged --name=calico-node -e HOSTNAME=$ADVERTISE_IP -e IP=$ADVERTISE_IP -e CALICO_DISABLE_FILE_LOGGING=true -e CALICO_NETWORKING=false -e ETCD_ENDPOINTS=$ETCD_ENDPOINTS -v /var/run/calico:/var/run/calico calico/node:latest
```
Comment (philips) [inline Documentation/deploy-workers.md]: any reason to not use rkt here?

File: multi-node/generic/worker-install.sh [hunk 2]
```diff
@@ -47,11 +47,22 @@ function init_templates {
...
 [Service]
-ExecStartPre=/usr/bin/mkdir -p /etc/kubernetes/manifests
-
 Environment=KUBELET_VERSION=${K8S_VER}
+Environment=HOSTNAME=${ADVERTISE_IP}
+Environment="RKT_OPTS=--volume cni,kind=host,source=/opt/cni/bin --mount volume=cni,target=/opt/cni/bin"
+ExecStartPre=/usr/bin/mkdir -p /etc/kubernetes/manifests
+ExecStartPre=/usr/bin/mkdir -p /opt/cni/bin
+ExecStartPre=/usr/bin/chmod a+w /opt/cni/bin
+ExecStartPre=-/usr/bin/wget -nc -O /opt/cni/bin/calico https://circle-artifacts.com/gh/projectcalico/calico-cni/135/artifacts/0/home/ubuntu/calico-cni/dist/calico
```
Comment (philips) [inline multi-node/generic/worker-install.sh]: Any way we can package these up into containers? If not can we document what the blockers are?

File: Documentation/deploy-master.md [hunk 3]
```diff
@@ -36,9 +36,10 @@ $ sudo chmod 600 /etc/kubernetes/ssl/*-key.pem
...
 
-### flannel Configuration
-
-[flannel][flannel-docs] provides a key Kubernetes networking capability &mdash; a software-defined overlay network to manage routing of the [Pod][pod-overview] network.
+### Network Configuration
+Networking is handled by two different components.
+* [flannel][flannel-docs] provides a key Kubernetes networking capability &mdash; a software-defined overlay network to manage routing of the [Pod][pod-overview] network.
+* [calico][calico-docs] provides additional security by adding network policy to the [Pod][pod-overview] network.
 
 *Note:* If the pod-network is being managed independently of flannel, this step can be skipped. See [kubernetes networking](kubernetes-networking.md) for more detail.
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: Presumably we'll need to rework this and the following paragraphs too?  e.g. May still want Calico policy if not running flannel such as GCE /24 type of a setup.

File: Documentation/deploy-master.md [hunk 4]
```diff
@@ -36,9 +36,10 @@ $ sudo chmod 600 /etc/kubernetes/ssl/*-key.pem
 $ sudo chown root:root /etc/kubernetes/ssl/*-key.pem
 ```
 
-### flannel Configuration
-
-[flannel][flannel-docs] provides a key Kubernetes networking capability &mdash; a software-defined overlay network to manage routing of the [Pod][pod-overview] network.
+### Network Configuration
+Networking is handled by two different components.
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: Networking is provided by Flannel and Calico.

File: Documentation/deploy-master.md [hunk 5]
```diff
@@ -36,9 +36,10 @@ $ sudo chmod 600 /etc/kubernetes/ssl/*-key.pem
...
+* [flannel][flannel-docs] provides a key Kubernetes networking capability &mdash; a software-defined overlay network to manage routing of the [Pod][pod-overview] network.
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: flannel provides a software-defined overlay network for routing traffic to/from the [Pod]s
Calico secures the overlay network by restricting traffic to/from the [Pod]s based on fine-grained [network policy]

File: Documentation/deploy-master.md [hunk 6]
```diff
@@ -63,26 +63,10 @@ Next create a [systemd drop-in][dropins], which is a method for appending or ove
 ExecStartPre=/usr/bin/ln -sf /etc/flannel/options.env /run/flannel/options.env
 ```
 
+[calico-docs]: https://docs.projectcalico.org
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: Until we consolidate our docs I think this needs to point to our Kubernetes docs on github.  We can more easily beef up the landing page there if required to point to docs outside of that repo.

File: Documentation/deploy-master.md [hunk 7]
```diff
@@ -364,6 +349,65 @@ spec:
       timeoutSeconds: 1
 ```
 
+### Run the per host calico agent
+
+The per host calico runs on all hosts including the master node. It is responsible for enforcing networking policy and programming routes into pods running on that host.
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: The per host Calico agent runs on...

File: Documentation/deploy-master.md [hunk 8]
```diff
@@ -364,6 +349,65 @@ spec:
...
+Restart=always
+TimeoutStartSec=0
+
+[Install]
+WantedBy=multi-user.target
+```
+
+### Set Up the policy-agent Pod
+
+The policy agent is the last major piece of the our master node. It monitors the API for changes related to network policy and configures Calico to implement that policy.
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: Is this actually running on the master node?  Or is it just wherever Kubernetes schedules it?

File: Documentation/deploy-master.md [hunk 9]
```diff
@@ -432,12 +491,38 @@ Now we can create the `kube-system` namespace:
...
 
+The Calico policy-agent runs in it's own `calico-system` namespace:
+
+```sh
+$ curl -H "Content-Type: application/json" -XPOST -d'{"apiVersion":"v1","kind":"Namespace","metadata":{"name":"kube-system"}}' "http://127.0.0.1:8080/api/v1/namespaces"
+```
+
+### Enable network policy API
+
+Network policy in Kubernetes is currently implemented as a 3rd party API. To enable it, run the following
```
Comment (lxpollitt) [inline Documentation/deploy-master.md]: 3rd party resource

File: Documentation/deploy-master.md (hunk 8 above)
Comment (tomdee) [inline Documentation/deploy-master.md]: Actually on the master

**Pull request closed without merging**